Subcategories (#861)
UniMath/agda-unimath: src/foundation/subsingleton-induction.lagda.md

# Subsingleton induction

```agda
module foundation.subsingleton-induction where
```

<details><summary>Imports</summary>

```agda
open import foundation.action-on-identifications-functions
open import foundation.dependent-pair-types
open import foundation.singleton-induction
open import foundation.universe-levels

open import foundation-core.contractible-types
open import foundation-core.function-types
open import foundation-core.homotopies
open import foundation-core.identity-types
open import foundation-core.propositions
open import foundation-core.sections
open import foundation-core.transport-along-identifications
```

</details>

## Idea

**Subsingleton induction** on a type `A` is a slight variant of
[singleton induction](foundation.singleton-induction.md) which in turn is a
principle analogous to induction for the [unit type](foundation.unit-type.md).
Subsingleton induction uses the observation that a type equipped with an element
is [contractible](foundation-core.contractible-types.md) if and only if it is a
[proposition](foundation-core.propositions.md).

> Subsingelton induction states that given a type family `B` over `A`, to
> construct a section of `B` it suffices to provide a section over `a` for some
> `a : A`.

Reviewer: ```suggestion
Subsingelton induction states that given a type family `B` over `A`, to
construct a section of `B` it suffices to construct an element of `B a`.
```

## Definition

### Subsingleton induction

```agda
is-subsingleton :
  (l1 : Level) {l2 : Level} (A : UU l2) → UU (lsuc l1 ⊔ l2)
is-subsingleton l A = (B : A → UU l) (a : A) → section (ev-point a {B})

ind-is-subsingleton :
  {l1 l2 : Level} {A : UU l1} →
  ({l : Level} → is-subsingleton l A) → (B : A → UU l2) (a : A) →
  B a → (x : A) → B x
ind-is-subsingleton is-subsingleton-A B a = pr1 (is-subsingleton-A B a)

compute-ind-is-subsingleton :
  {l1 l2 : Level} {A : UU l1} (H : {l : Level} → is-subsingleton l A) →
  (B : A → UU l2) (a : A) → (ev-point a {B} ∘ ind-is-subsingleton H B a) ~ id
compute-ind-is-subsingleton is-subsingleton-A B a = pr2 (is-subsingleton-A B a)
```

## Properties

### Propositions satisfy subsingleton induction

```agda
abstract
  ind-subsingleton :
    {l1 l2 : Level} {A : UU l1} (is-prop-A : is-prop A)

>     (B : A → UU l2) → (a : A) → B a → (x : A) → B x

Reviewer: I think it would be nice to make `a` implicit here. I saw also later on that you're just using an underscore when you apply singleton induction. Can we also make `B` implicit?

Reviewer: Probably `a` can be implicit in a bunch more places.

  ind-subsingleton is-prop-A B a =
    ind-singleton a (is-proof-irrelevant-is-prop is-prop-A a) B

abstract
  compute-ind-subsingleton :
    {l1 l2 : Level} {A : UU l1}
    (is-prop-A : is-prop A) (B : A → UU l2) (a : A) →
    (ev-point a {B} ∘ ind-subsingleton is-prop-A B a) ~ id
  compute-ind-subsingleton is-prop-A B a =
    compute-ind-singleton a (is-proof-irrelevant-is-prop is-prop-A a) B
```

> ### A type satisfies subsingleton induction if and only if it is a proposition

Reviewer: Beautiful!

# Subsingleton induction

```agda
module foundation.subsingleton-induction where
```

<details><summary>Imports</summary>

```agda
open import foundation.action-on-identifications-functions
open import foundation.dependent-pair-types
open import foundation.singleton-induction
open import foundation.universe-levels

open import foundation-core.contractible-types
open import foundation-core.function-types
open import foundation-core.homotopies
open import foundation-core.identity-types
open import foundation-core.propositions
open import foundation-core.sections
open import foundation-core.transport-along-identifications
```

</details>

## Idea

> **Subsingleton induction** on a type `A` equipped with an element `a : A` is a

Reviewer: The way I understand subsingleton induction is that it it just on a type, not on a type equipped with an element.

Author: yeah, sorry, I just fixed it
